// @11ty/font-awesome/src/get-prefixes.js
import fs from "node:fs";

async function getPrefixes() {
	let url = "https://api.fontawesome.com";

	let data = await fetch(url, {
		method: "post",
		body: JSON.stringify({
			"query": "query { release(version:\"6.x\"){ familyStyles { prefix family style } } }"
		}),
		headers: {
			"Content-Type": "application/json",
		},
	});

	let json = await data.json();

	let map = {};
	let styles = new Set();
	let families = new Set();

	for(let {family, prefix, style} of json?.data?.release?.familyStyles || []) {
		styles.add(style);
		families.add(family);

		if(!map[family]) {
			map[family] = {};
		}
		if(!map[family][style]) {
			map[family][style] = prefix;
		}
	}

	return {
		styles: Array.from(styles),
		families: Array.from(families),
		map,
	};
};

let prefixes = await getPrefixes();
let filePath = "./src/prefixes.js";
fs.writeFileSync(filePath, `// Warning: this file is automatically generated by \`npm run update-prefixes\`
export default ${JSON.stringify(prefixes, null, 2)}`);
console.log( `Prefixes written to ${filePath}.` );
